src/ray/raylet/scheduling_resources.cc

```diff
@@ -83,62 +134,43 @@ bool ResourceSet::IsEqual(const ResourceSet &rhs) const {
-    const double &resource_capacity = resource_pair.second;
-    if (resource_capacity_.count(resource_label) == 1) {
-      resource_capacity_[resource_label] -= resource_capacity;
-      if (resource_capacity_[resource_label] < 0 + EPSILON) {
-        resource_capacity_.erase(resource_label);
-      }
-    }
-  }
-}
 
 void ResourceSet::SubtractResourcesStrict(const ResourceSet &other) {
   // Subtract the resources, make sure none goes below zero and delete any if new capacity
   // is zero.
-  for (const auto &resource_pair : other.GetResourceMap()) {
+  for (const auto &resource_pair : other.GetResourceAmountMap()) {
     const std::string &resource_label = resource_pair.first;
-    const double &resource_capacity = resource_pair.second;
+    const FractionalResourceQuantity &resource_capacity = resource_pair.second;
     RAY_CHECK(resource_capacity_.count(resource_label) == 1)
         << "Attempt to acquire unknown resource: " << resource_label;
     resource_capacity_[resource_label] -= resource_capacity;
-    // TODO(romilb): Double precision subtraction may sometimes be less than zero by a
-    // small epsilon - need to fix.
-    RAY_CHECK(resource_capacity_[resource_label] >= 0 - EPSILON)
-        << "Capacity of resource " << resource_label << " after subtraction is negative ("
-        << resource_capacity_[resource_label] << ")."
-        << " Debug: resource_capacity_:" << ToString() << ", other: " << other.ToString();
-    if (EqualsZeroEpsilon(resource_capacity_[resource_label])) {
+    if (resource_capacity_[resource_label] == 0) {
       resource_capacity_.erase(resource_label);
     }
   }
 }
 
 // Perform an outer join.
 void ResourceSet::AddResources(const ResourceSet &other) {
-  for (const auto &resource_pair : other.GetResourceMap()) {
+  for (const auto &resource_pair : other.GetResourceAmountMap()) {
     const std::string &resource_label = resource_pair.first;
-    const double &resource_capacity = resource_pair.second;
+    const FractionalResourceQuantity &resource_capacity = resource_pair.second;
     resource_capacity_[resource_label] += resource_capacity;
   }
 }
 
-double ResourceSet::GetResource(const std::string &resource_name) const {
-  if (EqualsZeroEpsilon(resource_capacity_.count(resource_name))) {
+FractionalResourceQuantity ResourceSet::GetResource(
+    const std::string &resource_name) const {
+  if (resource_capacity_.count(resource_name) == 0) {
     return 0;
   }
-  double capacity = resource_capacity_.at(resource_name);
-  RAY_CHECK(capacity > 0 + EPSILON) << "Resource " << resource_name << " capacity is "
-                                    << capacity
-                                    << ". Should have been greater than zero.";
+  const FractionalResourceQuantity capacity = resource_capacity_.at(resource_name);
   return capacity;
 }
 
-double ResourceSet::GetNumCpus() const { return GetResource(kCPU_ResourceLabel); }
+double ResourceSet::GetNumCpus() const {
```

Comment:
I haven't looked into this closely, but I'm a little concerned that we may release some CPU resources and then return them and the returned quantity will be slightly different. Since we're converting to doubles in between.